src/org/mozilla/javascript/ScriptRuntime.java

```diff
@@ -2997,6 +3180,132 @@ public static CharSequence add(Object val1, CharSequence val2) {
+            if (val2.equals(BigInteger.ZERO)) {
+                throw ScriptRuntime.rangeErrorById("msg.division.zero");
+            }
+            return ((BigInteger)val1).remainder((BigInteger)val2);
+        } else if (val1 instanceof BigInteger || val2 instanceof BigInteger) {
+            throw ScriptRuntime.typeErrorById("msg.cant.convert.to.number", "BigInt");
+        } else {
+            return val1.doubleValue() % val2.doubleValue();
+        }
+    }
+
+    public static Number exponentiate(Number val1, Number val2) {
+        if (val1 instanceof BigInteger && val2 instanceof BigInteger) {
+            if (((BigInteger)val2).signum() == -1) {
+                throw ScriptRuntime.rangeErrorById("msg.bigint.negative.exponent");
+            }
+            // TODO: val2 is supported only in the range of 32-bit integer.
+            return ((BigInteger)val1).pow(val2.intValue());
+        } else if (val1 instanceof BigInteger || val2 instanceof BigInteger) {
+            throw ScriptRuntime.typeErrorById("msg.cant.convert.to.number", "BigInt");
+        } else {
+            return Math.pow(val1.doubleValue(), val2.doubleValue());
+        }
+    }
+
+    public static Number bitwiseAND(Number val1, Number val2) {
+        if (val1 instanceof BigInteger && val2 instanceof BigInteger) {
+            return ((BigInteger)val1).and((BigInteger)val2);
+        } else if (val1 instanceof BigInteger || val2 instanceof BigInteger) {
+            throw ScriptRuntime.typeErrorById("msg.cant.convert.to.number", "BigInt");
+        } else {
+            int result = toInt32(val1.doubleValue()) & toInt32(val2.doubleValue());
+            return Double.valueOf(result);
+        }
+    }
+
+    public static Number bitwiseOR(Number val1, Number val2) {
+        if (val1 instanceof BigInteger && val2 instanceof BigInteger) {
+            return ((BigInteger)val1).or((BigInteger)val2);
+        } else if (val1 instanceof BigInteger || val2 instanceof BigInteger) {
+            throw ScriptRuntime.typeErrorById("msg.cant.convert.to.number", "BigInt");
+        } else {
+            int result = toInt32(val1.doubleValue()) | toInt32(val2.doubleValue());
+            return Double.valueOf(result);
+        }
+    }
+
+    public static Number bitwiseXOR(Number val1, Number val2) {
+        if (val1 instanceof BigInteger && val2 instanceof BigInteger) {
+            return ((BigInteger)val1).xor((BigInteger)val2);
+        } else if (val1 instanceof BigInteger || val2 instanceof BigInteger) {
+            throw ScriptRuntime.typeErrorById("msg.cant.convert.to.number", "BigInt");
+        } else {
+            int result = toInt32(val1.doubleValue()) ^ toInt32(val2.doubleValue());
+            return Double.valueOf(result);
+        }
+    }
+
+    public static Number leftShift(Number val1, Number val2) {
+        if (val1 instanceof BigInteger && val2 instanceof BigInteger) {
+            // TODO: val2 is supported only in the range of 32-bit integer.
```

Comment:
Do we need to enter a bug or something to fix this TODO? What will happen now if val2 is out of range?

Reply:
https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#intValue--

>if this BigInteger is too big to fit in an int, only the low-order 32 bits are returned

In Chrome when I try `1n << 2147483647n` (java.lang.Integer.MAX_VALUE) I get a RangeError: Maximum BigInt size exceeded.

It looks like java also throws an error with numbers that large:
```
js> var b = new java.math.BigInteger(1)
js> b.shiftLeft(java.lang.Integer.MAX_VALUE)
org.mozilla.javascript.WrappedException: Wrapped java.lang.ArithmeticException: BigInteger would overflow supported range
```
If the BigInteger is larger than Ingeter.MAX_VALUE you can get some pretty wacky incorrect results:

```
js> new java.math.BigInteger(java.lang.Integer.MAX_VALUE).shiftLeft(1).intValue()
-2
```

It looks like you can use BigInteger.intValueExact() instead, which will throw an ArithmeticException if the number is too large to fit in an int.

Reply:
Can you change it to throw RangeError if the size is out of range? That sounds quite reasonable, and then we're done. I think that's a lot more preferable to letting Java exceptions leak out.

Reply:
I looked at it again, and have another suggestion.

I think you will want to check if `val2.signum() == -1`, return  `signedRightShift(val1, val2.negate())`. Else if either `val1.shiftLeft()` or `val2.intValueExact()` throw an ArithmeticException, rethrow as a RangeError.

Likewise, in signedRightShift, if `val2.signum() == -1`, return `leftShift(val1, val2.negate())`. Else if `val2.intValueExact()` throws an ArithmeticException, return BigInteger.ZERO.

Reply:
https://github.com/mozilla/rhino/pull/837/commits/13dcf61bd7eddd57d1d858b42cc7d5b370e42ba3
I did this because not only `intValueExact`, but also `shiftLeft`, etc. sometimes threw `ArithmeticException`.
 I'm checking for error patterns in `testsrc/jstests/es2020/bigint.js`, please let me know if I'm missing anything.